src/tpm2.rs

```diff
@@ -0,0 +1,957 @@
+                "ownerread|policywrite|ownerwrite",
+                owner_pw,
+                owner_pw
+            ),
+        };
+
+        self.run(cmd, None)?;
+        self.run(
+            format!(
+                "tpm2_nvwrite -x 0x1500018 -a 0x40000001 -P {} {}",
+                owner_pw, nvpath
+            ),
+            None,
+        )?;
+        Ok(())
+    }
+
+    // TPM quote functions
+    fn create_quote(
+        &mut self,
+        nonce: String,
+        data: String,
+        mut pcrmask: String,
+    ) -> Result<String, KeylimeTpmError> {
+        let quote_tf = NamedTempFile::new()?;
+        let sign_tf = NamedTempFile::new()?;
+        let pcr_tf = NamedTempFile::new()?;
+
+        if pcrmask.is_empty() {
+            pcrmask = EMPTYMASK.to_string();
+        }
+
+        if !(data.is_empty()) {
+            self.run(
+                format!("tpm2_pcrreset {}", common::TPM_DATA_PCR),
+                None,
+            )?;
+            self.extend_pcr(common::TPM_DATA_PCR, data)?;
+        }
+
+        let quote_path = self.temp_file_get_path(&quote_tf)?;
+        let sign_path = self.temp_file_get_path(&sign_tf)?;
+        let pcr_path = self.temp_file_get_path(&pcr_tf)?;
+
+        let cmd = match self.legacy_tools {
+            true => format!(
+                "tpm2_quote -k {} -L {}:{} -q {} -m {} -s {} -p {} -G {} -P {}",
+                hex::encode(self.get_tpmdata("aik_handle")),
+                self.hash_alg,
+                self.pcr_mask_to_list(pcrmask, self.hash_alg.to_string())?,
+                hex::encode(nonce),
+                quote_path,
+                sign_path,
+                pcr_path,
+                self.hash_alg,
+                self.get_tpmdata("aik_pw"),
+            ),
+            false => format!(
+                "tpm2_deluxequote -C {} -L {}:{} -q {} -m {} -s {} -p {} -G {} -P {}",
+                hex::encode(self.get_tpmdata("aik_handle")),
+                self.hash_alg,
+                self.pcr_mask_to_list(pcrmask, self.hash_alg.to_string())?,
+                hex::encode(nonce),
+                quote_path,
+                sign_path,
+                pcr_path,
+                self.hash_alg,
+                self.get_tpmdata("aik_pw")
+            ),
+        };
+
+        self.run(cmd, Some(vec![quote_path, sign_path, pcr_path]))
+            .and_then(|(_, quotes)| {
+                let mut quote_list = Vec::new();
+                for val in quotes.values() {
+                    quote_list
+                        .push(self.base64_zlib_encode(val.to_string())?);
+                }
+                Ok(quote_list.as_slice().join(":"))
+            })
+    }
+
+    pub fn create_deep_quote(
+        &self,
+        nonce: String,
+        data: String,
+        pcrmask: String,
+        vpcrmask: String,
+    ) -> Result<String, KeylimeTpmError> {
+        Err(KeylimeTpmError::new_tpm_rust_error(
+            "Deep quote in progress.",
```

Comment:
This... seems wrong?  Why even have this function?  Or are you saying it's not done, at which point: please finish it.

Reply:
i don't mind removing this as its going to be re-implemented anyway (Its part of the old vTPM code that was prototyped on Xen). We plan to port this to QEMU, but the design is still in flight.